File: string/sol/isomorphic-strings.cpp
// TC = O(N)
// extra SC = O(2*256)
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if(s.size() != t.size()) return false;
        int n = s.size();
        vector<int> mp1(255 , -1);
        vector<int> mp2(255 , -1); 
        for(int i = 0; i< n;i++){
            char u = s[i];
            char v = t[i];
            // map u -> v. 
            if(mp1[u] != -1 && mp1[u] != v) return false;
            if(mp2[v] != -1 && mp2[v] != u) return false;
            
            mp1[u] = v;
            mp2[v] = u;
        }
        return true;
    }
};

File: string/sol/isomorphic-strings.java
// TC = O(N)
// extra SC = O(2*256)

// we don't allow many-to-one  or  one-to-many.
// we have to  maintain two maps.
// one for u -> v ... another for v -> u
class Solution {
    public boolean isIsomorphic(String s, String t) {
        HashMap<Character,Character> map = new HashMap<>();
        HashMap<Character,Character> revMap = new HashMap<>();
        int n = s.length();
        for(int i = 0;i < n;i++){
            char u = s.charAt(i);
            char v = t.charAt(i);
            if(map.containsKey(u) && map.get(u) != v) return false;
            if(revMap.containsKey(v) && revMap.get(v) != u) return false;
            map.put(u,v);
            revMap.put(v,u);
        }
        return true;
    }
}
